extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/BigQueryServiceImpl.java

```diff
@@ -0,0 +1,594 @@
+            return;
+        }
+
+        initCredentials();
+
+        var project = target.project();
+        if (project == null) {
+            project = gcpConfiguration.getProjectId();
+        }
+
+        var bqBuilder = BigQueryOptions.newBuilder().setProjectId(project);
+
+        if (credentials != null) {
+            bqBuilder.setCredentials(credentials);
+        }
+
+        bigQuery = bqBuilder.build().getService();
+
+        var settingsBuilder = BigQueryWriteSettings.newBuilder();
+        if (credentials != null) {
+            settingsBuilder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+        }
+
+        writeClient = BigQueryWriteClient.create(settingsBuilder.build());
+    }
+
+    private JSONObject buildRecord(FieldList fields, FieldValueList values) {
+        var record = new JSONObject();
+        int colCount = fields.size();
+        for (int i = 0; i < colCount; i++) {
+            record.put(fields.get(i).getName(), values.get(i).getValue());
+        }
+
+        return record;
+    }
+
+    private void refreshBigQueryCredentials() {
+        var credentials = bigQuery.getOptions().getCredentials();
+        if (!(credentials instanceof OAuth2Credentials)) {
+            return;
+        }
+
+        try {
+            // TODO check margin and refresh if too close to expire.
+            ((OAuth2Credentials) credentials).refreshIfExpired();
+        } catch (IOException ioException) {
+            monitor.warning("BigQuery Service cannot refresh credentials", ioException);
+        }
+    }
+
+    private void outputPage(OutputStream outputStream, TableResult tableResult) throws IOException {
+        var page = new JSONArray();
+
+        tableResult.getValues()
+                .forEach(row -> page.put(buildRecord(tableResult.getSchema().getFields(), row)));
+
+        outputStream.write(page.toString().getBytes());
+    }
+
+    private void setNamedParameters(QueryJobConfiguration.Builder queryConfigBuilder, DataAddress sinkAddress) {
+        var prefix = EDC_NAMESPACE + '@';
+        for (var key : sinkAddress.getProperties().keySet()) {
+            if (key.startsWith(prefix)) {
+                key = key.substring(EDC_NAMESPACE.length());
+
+                var value = sinkAddress.getStringProperty(key);
+                key = key.substring(1);
+                int separatorIndex = key.indexOf("_");
+                if (separatorIndex != -1) {
+                    var type = key.substring(0, separatorIndex);
+                    key = key.substring(separatorIndex + 1);
+                    var sqlValue = QueryParameterValue.of(value, StandardSQLTypeName.valueOf(type));
+                    queryConfigBuilder.addNamedParameter(key, sqlValue);
+                }
+            }
+        }
+    }
+
+    @Override
+    public Stream<DataSource.Part> runSourceQuery(String query, DataAddress sinkAddress) throws InterruptedException {
```

Comment:
Looks like this class is quite big and hard to read, I would split it in `read` and `write`, because they will be used separately anyway.

Reply:
Split into 3 classes, for provision, source and sink